amethyst_animation/src/systems/control.rs

```diff
@@ -0,0 +1,329 @@
+use std::time::{Duration, Instant};
+
+use amethyst_assets::AssetStorage;
+use amethyst_core::LocalTransform;
+use specs::{Entities, Fetch, Join, ReadStorage, System, WriteStorage};
+
+use resources::{Animation, AnimationCommand, AnimationControl, AnimationHierarchy,
+                AnimationOutput, ControlState, RestState, Sampler, SamplerControl,
+                SamplerControlSet};
+
+/// System for setting up animations, should run before `SamplerInterpolationSystem`.
+#[derive(Default)]
+pub struct AnimationControlSystem;
+
+impl AnimationControlSystem {
+    pub fn new() -> Self {
+        Default::default()
+    }
+}
+
+impl<'a> System<'a> for AnimationControlSystem {
+    type SystemData = (
+        Entities<'a>,
+        Fetch<'a, AssetStorage<Animation>>,
+        Fetch<'a, AssetStorage<Sampler>>,
+        WriteStorage<'a, AnimationControl>,
+        WriteStorage<'a, SamplerControlSet>,
+        ReadStorage<'a, AnimationHierarchy>,
+        ReadStorage<'a, LocalTransform>,
+    );
+
+    fn run(
+        &mut self,
+        (entities, animation_storage, sampler_storage, mut controls, mut samplers, hierarchies,
+            transforms): Self::SystemData,
+){
+        let now = Instant::now();
+        let mut remove = Vec::default();
+        for (entity, control, hierarchy) in (&*entities, &mut controls, &hierarchies).join() {
+            let new_state = match animation_storage.get(&control.animation) {
```

Comment:
Could we add a helper function for the `match` logic below? That way, we could call `animation_storage.get` and do `and_then` while keeping a flat layout.